src/protocol/sort/shuffle.rs

```diff
@@ -103,70 +119,121 @@ impl<'a, F: Field, S: Step + SpaceIndex> Shuffle<'a, F, S> {
+        permutations: &(Permutation, Permutation),
     ) -> Result<Vec<Replicated<F>>, BoxError> {
         let to_helper = Self::shuffle_for_helper(which_step);
         let step = (self.step_fn)(which_step);
-        let prss = &ctx.participant[step];
         let channel = ctx.gateway.get_channel(step);
 
         if to_helper != channel.identity() {
-            let direction = if to_helper.peer(Direction::Left) == channel.identity() {
-                Direction::Left
+            let permute = if to_helper.peer(Direction::Left) == channel.identity() {
+                &permutations.0
             } else {
-                Direction::Right
+                &permutations.1
             };
-            let mut permute = Self::generate_random_permutation(self.input.len(), direction, prss);
-            apply_inv(&mut permute, &mut self.input);
+
+            if matches!(shuffle_or_unshuffle, ShuffleOrUnshuffle::Shuffle) {
+                apply_inv(&mut permute.clone(), &mut self.input);
+            } else {
+                apply(&mut permute.clone(), &mut self.input);
+            }
         }
         self.reshare_all_shares(ctx, which_step).await
     }
+
     #[embed_doc_image("shuffle", "images/sort/shuffle.png")]
-    /// Shuffle calls `single_shuffle` three times with 2 helpers shuffling the shares each time.
-    /// Order of calling `single_shuffle` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
+    /// Shuffle calls `single` three times with 2 helpers shuffling the shares each time.
+    /// Order of calling `single` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
     /// Each single shuffle requires communication between helpers to perform reshare.
     /// Infrastructure has a pre-requisite to distinguish each communication step uniquely.
-    /// For this, we have three shuffle steps one per `single_shuffle` i.e. Step1, Step2 and Step3.
+    /// For this, we have three shuffle steps one per `single` i.e. Step1, Step2 and Step3.
     /// The Shuffle object receives a step function and appends a `ShuffleStep` to form a concrete step
     /// ![Shuffle steps][shuffle]
     #[allow(dead_code)]
-    pub async fn execute<M: Mesh, G: Gateway<M, S>>(
+    pub async fn execute_shuffle<M: Mesh, G: Gateway<M, S>>(
+        &mut self,
+        ctx: &ProtocolContext<'_, G, S>,
+        permutations: &(Permutation, Permutation),
+    ) -> Result<(), BoxError>
+    where
+        F: Field,
+    {
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Shuffle, ctx, Step1, permutations)
+            .await
+            .unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Shuffle, ctx, Step2, permutations)
+            .await
+            .unwrap();
+        *self.input = self
+            .single(ShuffleOrUnshuffle::Shuffle, ctx, Step3, permutations)
+            .await
+            .unwrap();
+
+        Ok(())
+    }
+
+    #[embed_doc_image("unshuffle", "images/sort/unshuffle.png")]
```

Comment:
This unshuffle diagram confused me more. I'm not sure why only one helper needs to "get permutation" each step. Isn't it two?

Reply:
made it better based on suggestions